design/20190708.certificate-request-crd.md

```diff
@@ -174,6 +188,190 @@ type ObjectReference struct {
+
+```
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: my-example-io-my-issuer-myapp-approver
+rules:
+  - apiGroups: ["cert-manager.io"]
+    resources: ["signers"]
+    verbs: ["approve"]
+    resourceNames: ["myissuers.my-example.io/myapp"]
+```
+
+These permissions are enforced using a
+[SubjectAccessReview](https://github.com/kubernetes/kubernetes/blob/b11d0fbdd58394a62622787b38e98a620df82750/pkg/apis/authorization/types.go#L27)
+that is performed in the cert-manager Webhook component when a user attempts to
+set the Approve or Denied condition. The user of the approver is tested along
+with the IssuerRef of the CertificateRequest.
+
+If the approver does not have sufficient permissions defined above to set the
+Approved or Denied conditions, the request will be rejected by the cert-manager
+validating admission webhook.
+
+###### Scope
+The RBAC permissions *must* be granted at the cluster scope (i.e.
+ClusterRoleBinding). This matches the same scoping behaviour as
+[certificates.k8s.io
+CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).
+
+Namespaced signers are represented by a namespaced resource using the syntax of
+`<signer-resource-name>.<signer-group>/<signer-namespace>.<signer-name>`
+
+###### API Group
+The apiGroup must *always* be `cert-manager.io`, as all CertificateRequests are
+in the `cert-manager.io` group. Even though some signers may themselves not
+belong to the `cert-manager.io` group, the approve operation does.
+
+###### Resource
+The resource must *always* be `signers`. This matches the same resource as
+[required by certificates.k8s.io
+CertificateSigningRequests](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#authorization).
+
+###### Verbs
+The verb must *always* be `approve`. This verb grants the approver the
+permissions to set *both* Approved and Denied conditions.
+
+###### Resource Names
+The resource name is made up of 3 parts: the signer resource name, signer group, and the
+signer name.
+
+The signer resource name should be the same string of the resource name of the
+signer.
+
+The signer group should be the same string of the group of signer.
+
+The name may be either a wild card '*' denoting all signer names in all
+namespaces, or instead a string which must match optionally a namespace,
+followed by the signer name as it appears in the IssuerRef.
+
+An example of signing all `myissuer` signers in all namespaces, and
+`clustermyissuers` with the name `myppp`, in the `my-example.io` group:
+
+```yaml
+    resourceNames: ["myissuers.my-example.io/*", "clustermyissuers.my-example.io/myapp"]
+```
+
+An example of signing `myissuer` with the name `myapp` in the namesapces `foo`
+and `bar`:
+
+```yaml
+    resourceNames: ["myissuers.my-example.io/foo.myapp", "myissuers.my-example.io/bar.myapp"]
+```
```

Comment:
Suggestion: the reader might find it easier to understand with some comments:

~~~suggestion
```yaml
resourceNames:
  # With the following, the user (i.e., the approver) can only approve the CertificateRequests
  # in the namespace "foo" and that are using the namespace-scoped MyIssuer named
  # "myapp":
  - myissuers.my-example.io/foo.myapp       
  # Similar to the above, but in the namespace "bar":
  - myissuers.my-example.io/bar.myapp
```
~~~